### ISSUE TITLE
fasta: Support for metadata annotations on other objects than functions/classes

### ISSUE THREAD

**Issue by mkustermann** [dart-lang/sdk#28434]
There are a number of tests which test that it is possible to have metadata annotations on other objects than functions/classes, see `co19/Language/Metadata/*`.

/cc @kmillikin & #28263

**Comment by mraleph** [comment]
Issue is still valid for fasta.

**Comment by peter-ahe-google** [comment]
Blocked on #29169

**Comment by scheglov** [comment]
BTW, it's not just these blocked nodes, metadata is also allowed on local variable declarations and variable declarations in `for` loops.

**Comment by peter-ahe-google** [comment]
I've added "Annotations on VariableDeclaration" to #29169.

**Comment by scheglov** [comment]
Why is it blocked?
Nothing blocks you from implementing it in the parser.
Kernel is not the only client.
There are failing tests in analyzer that require parser support for these annotations.

**Comment by peter-ahe-google** [comment]
@scheglov This is blocked because all the failures the OP mentioned require changes to kernel AST.

But I can take a look at the missing parser support. Can you point me to the failing tests?

**Comment by scheglov** [comment]
OK, I see.
Still, there are more than one client that needs to pass these tests.

There are more tests in analyzer, but most of them have not been moved yet to verify with Fasta.
[These two](https://github.com/dart-lang/sdk/blob/5fcbc900a6ad67d6b84f21d3f5a4cb7c4be93e0f/pkg/analyzer/test/generated/parser_fasta_test.dart#L904) have.

**Comment by peter-ahe-google** [comment]
I got confused, this is still partially blocked, and it is not metadata in general (which @stereotype441 is working on).

**Comment by scheglov** [comment]
Is this still blocked?
https://github.com/dart-lang/sdk/issues/30035 was fixed, it is not possible to represent annotations of local variables and parameters.

**Comment by peter-ahe-google** [comment]
Thank you, @scheglov. This is no longer blocked.

**Comment by peter-ahe-google** [comment]
Let me check.

**Comment by peter-ahe-google** [comment]
@danrubel @bwilkerson

**Comment by peter-ahe-google** [comment]
@vsmenon is it only annotations on parameters you're missing?

**Comment by vsmenon** [comment]
That's the only thing we're aware of at the moment.

@jmesserly - do you see anything else missing?

**Comment by jmesserly** [comment]
Yeah, parameters are the only blocking thing that I'm aware of.

Local variable annotations would be nice to have, but they aren't required/urgent. (DDC uses some local variable annotations in its private SDK impl code, to indicate non-null locals. But that can be refactored to use something else.)

**Comment by danrubel** [comment]
The fasta parser has supported metadata on formal method parameters for a while and it is plumbed through the AstBuilder to Analyzer. The BodyBuilder appears to ignore metadata on formal parameters.

@jmesserly I'm not familiar with how DDC hooks into fasta. Is this what you need?

**Comment by jmesserly** [comment]
> @jmesserly I'm not familiar with how DDC hooks into fasta. Is this what you need?

DDC consumes the Kernel API. I'm not seeing parameter annotations filled in for the Kernel tree. Maybe it's BodyBuilder that hasn't implemented them yet?

**Comment by jmesserly** [comment]
This patch appears to fix it, and the metadata shows up in the Kernel AST for DDC to use. But it may be incorrect because I haven't looked at this code before.

```diff
--- a/pkg/front_end/lib/src/fasta/kernel/body_builder.dart
+++ b/pkg/front_end/lib/src/fasta/kernel/body_builder.dart
@@ -2096,7 +2096,7 @@ class BodyBuilder<Arguments> extends ScopeListener<JumpTarget>
     }
     bool isConst = (modifiers & constMask) != 0;
     bool isFinal = (modifiers & finalMask) != 0;
-    ignore(Unhandled.Metadata);
+    List annotations = pop();
     VariableDeclaration variable;
     if (!inCatchClause &&
         functionNestingLevel == 0 &&
@@ -2124,6 +2124,12 @@ class BodyBuilder<Arguments> extends ScopeListener<JumpTarget>
         variable.fileOffset = offsetForToken(name.token);
       }
     }
+    if (annotations != null) {
+      _typeInferrer.inferMetadata(this, annotations);
+      for (Expression annotation in annotations) {
+        variable.addAnnotation(annotation);
+      }
+    }
     push(variable);
   }
```

**Comment by peter-ahe-google** [comment]
Thank you, @jmesserly. @stefantsov will take this suggestion and add test cases.

**Comment by stefantsov** [comment]
I believe the issue is about to be solved in [CL 51840](https://dart-review.googlesource.com/c/sdk/+/51840).  I added a test case and updated a couple of existing ones.  The most tricky part was to properly invoke the inference, so that it doesn't work on the same node twice.  I don't expect serious issues with the CL, and I think it's ready for the review.

In addition to variable declarations (that represent both variables and formal parameters), I handled the annotations on local functions.  Each local function declaration already contains a variable declaration that represents this function declaration and that can hold the annotations.

While working on this I also looked into annotations for type parameters.  They are now handled if they are declared in local functions or function expressions.  I hope to get to type parameters of classes, typedefs, and top-level functions soon.  Until then I'm using references to https://github.com/dart-lang/sdk/issues/28981 in status files.

**Comment by stefantsov** [comment]
The support for annotations on formals, local variables, and local function declarations is added in https://github.com/dart-lang/sdk/commit/0289071afed95040676bd71d9e67e5d6ab688bbd.

**Comment by stefantsov** [comment]
https://github.com/dart-lang/sdk/commit/0289071afed95040676bd71d9e67e5d6ab688bbd had to be reverted :-(  For some reason, the precomp bots aren't happy.  The following test breaks:

```
python tools/test.py -m release -c dartkp -r dart_precompiled --strong language_2/metadata_test
```

Also, Flutter won't build :-(  I'll investigate the reason next week and hope to get the patch back soon.

**Comment by kmillikin** [comment]
As far as I know we have annotations everywhere now.  Reopen this issue if you discover that's not the case.
